### PR TITLE
initial port to python v3

### PR DESCRIPTION
**Pull request by jordic**

### CODE REVIEW

**Comment by jordic** [comment]
still lacks tests

**Comment by icemac** [comment]
I converted this PR into a draft as currently the changes even break the tests on Python 2. (See TravisCI.) It seems at least a `from __future__ import print_function` is needed.

**Comment by jordic** [comment]
> break the tests on Python 2. (See TravisCI.) I

does make sense to keep python 2 for a new version?

**Comment by dataflake** [comment]
We will only remove Python 2 support if there's a good technical reason because Zope 4 still supports Python 2.

File: src/z3c/sqlalchemy/base.py [hunk 1]
```diff
@@ -137,5 +136,4 @@ def _createEngine(self):
                                             autocommit=not self.transactional,
                                             twophase=self.twophase,
                                             autoflush=True,
-                                            extension=ZopeTransactionExtension(**self.extension_options),
```
Comment (mgedmin) [inline src/z3c/sqlalchemy/base.py:140]: This is unrelated to the Python 3 port and should be a part of a separate PR.

File: src/z3c/sqlalchemy/model.py [hunk 2]
```diff
@@ -8,32 +8,31 @@
...
-    implements(IModel)
+    """
 
     def __init__(self, *args):
         """ The constructor can be called with a series of dict. Each dict
             represents a single table and its data (see add() method).
         """
 
-        super(Model, self).__init__()
+        super().__init__()
```
Comment (mgedmin) [inline src/z3c/sqlalchemy/model.py:35]: This drops Python 2 compatibility.

I'd prefer if we had at least one version working with both Python 2 and Python 3.

File: src/z3c/sqlalchemy/postgres.py [hunk 3]
```diff
@@ -12,44 +12,43 @@
...
-            for row in rs: 
+            for row in rs:
                 tablename = row.tablename
 
                 try:
                     table = sqlalchemy.Table(tablename, db, autoload=True)
                 except KeyError:
                     if ignoreErrors:
-                        print >>sys.stderr, 'Can\'t load table %s' % tablename
+                        print('Can\'t load table %s' % tablename, file=sys.stderr)
```
Comment (mgedmin) [inline src/z3c/sqlalchemy/postgres.py:47]: This needs a `from __future__ import print_function` at the top.

File: src/z3c/sqlalchemy/test.py [hunk 4]
```diff
@@ -47,6 +47,4 @@ class HierarchyNode(MappedClassBase):
 
 H = wrapper.getMapper('hierarchy')
 
-print H
 rows = session.query(H).select_by(H.c.id==8)
-print rows[0].children
```
Comment (mgedmin) [inline src/z3c/sqlalchemy/test.py:52]: I think this entire file needs to go.

**Review by mgedmin — CHANGES_REQUESTED** [review]
Mostly good!  Some changes needed

File: src/z3c/sqlalchemy/base.py (hunk 1 above)
Comment (jordic) [reply src/z3c/sqlalchemy/base.py:140]: Got it, I just removed it because it was failing, anyway here's the path
https://github.com/zopefoundation/zope.sqlalchemy/blob/master/CHANGES.rst#12-2019-10-17

**Comment by dataflake** [comment]
You have now mixed a whole bunch of cosmetic changes into this PR by running `black` over it. This is an absolute no-go.

File: .travis.yml [hunk 5]
```diff
@@ -1,10 +1,22 @@
+# Wire up travis
 language: python
-python:
-    - 2.7
+sudo: false
```
Comment (dataflake) [inline .travis.yml:3]: This can go

File: .travis.yml [hunk 6]
```diff
@@ -1,10 +1,22 @@
...
+      env: TOXENV=py36
+    - python: 3.7
+      env: TOXENV=py37
+    - python: 3.8
+      env: TOXENV=py38
+
 install:
-    - python bootstrap.py
-    - bin/buildout
+  - travis_retry pip install tox
```
Comment (dataflake) [inline .travis.yml:17]: `travis_retry` is not needed

File: .travis.yml [hunk 7]
```diff
@@ -1,10 +1,22 @@
...
+
 script:
-    - bin/test -v1
+  - travis_retry tox
```
Comment (dataflake) [inline .travis.yml:20]: see above

File: src/z3c/sqlalchemy/__init__.py [hunk 8]
```diff
@@ -7,10 +7,10 @@
 ##########################################################################
 
 
-from util import *
-from model import Model
+from z3c.sqlalchemy.util import *
+from z3c.sqlalchemy.model import Model
```
Comment (dataflake) [inline src/z3c/sqlalchemy/__init__.py:12]: I'd use a relative import instead

File: src/z3c/sqlalchemy/model.py [hunk 9]
```diff
@@ -8,29 +8,28 @@
...
 """
 
 import sqlalchemy
-from zope.interface import implements
+from zope.interface import implementer
 
-from mapper import MappedClassBase
-from interfaces import IModel
+from z3c.sqlalchemy.mapper import MappedClassBase
+from z3c.sqlalchemy.interfaces import IModel
```
Comment (dataflake) [inline src/z3c/sqlalchemy/model.py:18]: I'd use a relative import instead

File: src/z3c/sqlalchemy/postgres.py [hunk 10]
```diff
@@ -12,52 +14,51 @@
 
 import sqlalchemy
 
-from zope.interface import implements
+from zope.interface import implementer
 
 from z3c.sqlalchemy.interfaces import ISQLAlchemyWrapper
 from z3c.sqlalchemy.base import ZopeWrapper
```
Comment (dataflake) [inline src/z3c/sqlalchemy/postgres.py:20]: I'd use a relative import instead

File: tox.ini [hunk 11]
```diff
@@ -0,0 +1,21 @@
+[tox]
+envlist = py27,py37,py38
+
+[testenv]
+sitepackages=False
+basepython =
+    {py27}: {env:TOXPYTHON:python2.7}
+    {py37}: {env:TOXPYTHON:python3.7}
+    {py38}: {env:TOXPYTHON:python3.8}
```
Comment (dataflake) [inline tox.ini:9]: Not needed

**Review by dataflake — CHANGES_REQUESTED** [review]
Leave out the `black` cosmetic changes. You also seem to be missing a suitable `.gitignore` - this PR is adding `.pyc` files.

**Comment by dataflake** [comment]
You keep going against conventions found in the vast majority of packages in the `zopfoundation` organization. We don't use black. We put configurations (e.g. isort) into setup.cfg. Would you please take a look at existing packages like https://github.com/zopefoundation/Zope before continuing?

**Comment by jordic** [comment]
> You keep going against conventions found in the vast majority of packages in the `zopfoundation` organization. We don't use black. We put configurations (e.g. isort) into setup.cfg. Would you please take a look at existing packages like https://github.com/zopefoundation/Zope before continuing?

Ohhhh, so sorry :( 
Wait a few, will try to do whatever you are asking :) happy to follow zopefoundation conventions :)

**Comment by jordic** [comment]
It's still wip, I think we also should update setup.py to reflect new deps. Anyway, just tell me whatever you want. 
Anyway, we use this package on a 2.7 zope install, and I'm also planning to use it on a new zope5 install.

File: .gitignore [hunk 12]
```diff
@@ -0,0 +1,13 @@
...
+env/
+venv/
+.env
+.idea
+.pytest_cache/
+.mypy_cache/
+bin/
+lib/
+lib64
+**/$*.py
```
Comment (mgedmin) [inline .gitignore:11]: Out of curiosity, what is this?

File: .gitignore [hunk 13]
```diff
@@ -0,0 +1,13 @@
...
+apiadmin.egg-info/
```
Comment (mgedmin) [inline .gitignore:12]: I think you meant to add '*.egg-info/' but it got expanded by your shell to some irrelevant project.

File: .travis.yml [hunk 14]
```diff
@@ -1,10 +1,21 @@
...
+matrix:
+  include:
+    - python: 2.7
+      env: TOXENV=py27
+    - python: 3.6
+      env: TOXENV=py36
+    - python: 3.7
+      env: TOXENV=py37
+    - python: 3.8
+      env: TOXENV=py38
```
Comment (mgedmin) [inline .travis.yml:13]: I would suggest [tox-travis](https://pypi.org/project/tox-travis/) to avoid repetition.

(The main reason we don't normally use tox-travis in other zopefoundation repos is because we want to measure test coverage and upload it to coveralls in our .travis.yml, but not in tox.ini, but AFAICT this is not currently set up for z3c.sqlalchemy.)

File: src/z3c/sqlalchemy/base.py (hunk 1 above)
Comment (mgedmin) [reply src/z3c/sqlalchemy/base.py:140]: It is not enough to remove ZopeTransactionExtension, you also have to call zope.sqlalchemy.register().

A separate PR that does just that would be nice.

File: src/z3c/sqlalchemy/postgres.py [hunk 15]
```diff
@@ -1,3 +1,4 @@
+from __future__ import print_function
```
Comment (mgedmin) [inline src/z3c/sqlalchemy/postgres.py:1]: Heh, when I said "at the top", I meant "right after the block comment".

(It's minor.)

File: src/z3c/sqlalchemy/postgres.py [hunk 16]
```diff
@@ -12,52 +13,51 @@
...
+
                 for c in table.c:
                     fk = c.foreign_key
                     if fk is not None:
 
                         ref_by_table = fk.column.table
                         ref_by_table_name = ref_by_table.name
 
-                        if not d.has_key(ref_by_table_name):
+                        if not ref_by_table_name in d:
```
Comment (mgedmin) [inline src/z3c/sqlalchemy/postgres.py:60]:
```suggestion
                        if ref_by_table_name not in d:
```

(minor)

File: tox.ini [hunk 17]
```diff
@@ -0,0 +1,17 @@
+[tox]
+envlist = py27,py36,py37,py38
+
+[testenv]
+sitepackages=False
```
Comment (mgedmin) [inline tox.ini:5]: That's the default, it's not necessary to repeat it.

File: tox.ini [hunk 18]
```diff
@@ -0,0 +1,17 @@
...
+deps = pytest               # PYPI package providing pytest
+commands = python -m py.test tests
+passenv = *
+extras =
+    testing
```
Comment (mgedmin) [inline tox.ini:10]: There's no `testing` extra defined in setup.py , there's only `test`.

File: tox.ini [hunk 19]
```diff
@@ -0,0 +1,17 @@
...
+
+[testenv:py37]
+setenv =
+    PYTHONDONTWRITEBYTECODE=1
```
Comment (mgedmin) [inline tox.ini:14]: Why?

File: tox.ini [hunk 20]
```diff
@@ -0,0 +1,17 @@
+[tox]
+envlist = py27,py36,py37,py38
+
+[testenv]
+sitepackages=False
+deps = pytest               # PYPI package providing pytest
+commands = python -m py.test tests
+passenv = *
```
Comment (mgedmin) [inline tox.ini:8]: Why do we need this?

File: tox.ini [hunk 21]
```diff
@@ -0,0 +1,17 @@
+[tox]
+envlist = py27,py36,py37,py38
+
+[testenv]
+sitepackages=False
+deps = pytest               # PYPI package providing pytest
+commands = python -m py.test tests
```
Comment (mgedmin) [inline tox.ini:7]: Why not
```suggestion
commands = pytest tests
```

File: tox.ini [hunk 22]
```diff
@@ -0,0 +1,17 @@
+[tox]
+envlist = py27,py36,py37,py38
+
+[testenv]
+sitepackages=False
+deps = pytest               # PYPI package providing pytest
```
Comment (mgedmin) [inline tox.ini:6]: This should already be pulled in by the `test` extra.

File: setup.py [hunk 23]
```diff
@@ -9,6 +9,7 @@
 
 from setuptools import setup, find_packages
 
+
 CLASSIFIERS = [
     'Development Status :: 5 - Production/Stable',
     'Intended Audience :: Developers',
```
Comment (mgedmin) [inline setup.py:15]: Could you please add PyPI classifiers for supported Python versions after the `Programming Language :: Python` one?

```
'Programming Language :: Python :: 2.7',
'Programming Language :: Python :: 3.6',
'Programming Language :: Python :: 3.7',
'Programming Language :: Python :: 3.8',
'Programming Language :: Python :: 3.9',
```

**Review by mgedmin — APPROVED** [review]
Mostly LGTM; I'm still not happy about having one half of the ZopeTransactionExtension -> register() API change intermixed with Python 3 support changes.

I'd also like to request a changelog note in CHANGES.rst, saying "Added Python 3.6-3.9 support".

(And while we can bikeshed this until it's perfect, I think it would improve the current state of affairs even if we merged this as-is.)

File: .gitignore (hunk 12 above)
Comment (jordic) [reply .gitignore:11]: ohhhh so soryyy, take another one from another project, just cleaning it O_O. Sometimes speed is not a good friend :)

Comment (mgedmin) [reply .gitignore:11]: No, but what IDE/editor/thing creates files named like this?  I'm curious.

File: tox.ini (hunk 19 above)
Comment (jordic) [reply tox.ini:14]: sorry, that's try and error, till I moved tests outside package, seems like pytest & tox doesn't like it. And found this at stackoverflow, that obviously don't worked.

File: src/z3c/sqlalchemy/postgres.py (hunk 15 above)
Comment (jordic) [reply src/z3c/sqlalchemy/postgres.py:1]: Here we go!
https://github.com/zopefoundation/z3c.sqlalchemy/pull/11

**Comment by dataflake** [comment]
Please do not remove `buildout.cfg`. `zc.buildout` is the default installer for almost all zopefoundation packages and even though it's not strictly needed here it should remain as an obvious way to install and run a test runner.

File: tox.ini (hunk 19 above)
Comment (mgedmin) [reply tox.ini:14]: Whoa, I didn't notice that you moved the tests outside the package.

I don't like that; other zopefoundation packages conventionally use zope.testrunner and embed the tests inside the 'tests' subpackage of every package.  I thought z3c.alchemy was special.  I didn't want to insist on bringing it inline with the rest of ZF, but I am against bringing it out of line.

Comment (mgedmin) [reply tox.ini:14]: If you need a working example, we have plenty of ZF packages using zope.testrunner from tox.  Here's how [zope.sqlalchemy does that](https://github.com/zopefoundation/zope.sqlalchemy/blob/053587de5c811aeefcba890148dbcb1aa20f1950/tox.ini).

Comment (dataflake) [reply tox.ini:14]: @jordic in general could you please respect conventions you found in this package instead of making large changes to code style, package structure etc? Only change what's needed.

Comment (jordic) [reply tox.ini:14]: Sorry, I agree with both of you, I acted like if I were the owner of the package and applied what I'm used to do (without too much thinking that this is a package that's inside an organization, and the best is to follow the guidelans of it). 

On another side, I will pretty appreciate if you can point me on how to install a package like this using buildout, and being able to run test using zope.testrunner. I just arrived to the zope world too late, but I will apreciate a lot your effort helping me learn all of this tools and conventions. So, many thanks for your time reviewing and taking care of this :)

Comment (jordic) [reply tox.ini:14]: Found it, the key is to just launch `python bootstrap.py`. 
```
https://github.com/zopefoundation/BTrees/blob/master/bootstrap.py#L83-L86
```

Comment (jordic) [reply tox.ini:14]: This line on the original doesn't work. The script at http://peak.telecommunity.com/dist/ez_setup.py. Doesn't seem to be py3 friendly. Any hints on fixing this? 
https://github.com/zopefoundation/z3c.sqlalchemy/blob/master/bootstrap.py#L28-L30

Comment (mgedmin) [reply tox.ini:14]: I wouldn't worry about buildout too much, myself.  I've been adding tox.ini's to any ZF package I work on, because otherwise running tests is too painful.

python bootstrap.py has bitrotted quite badly in recent times and people are now recommending to install zc.buildout in a virtualenv somewhere and to run `buildout bootstrap` instead.  People are also removing bootstrap.py files from ZF repos.  buildout.cfg is usually left alone, let people who like buildout use it if they want to.

Comment (dataflake) [reply tox.ini:14]: @mgedmin I totally disagree. buildout is still the standard.

bootstrap.py will only work under Python 2 and its use is deprecated. To work on a package and test we now use virtual environments and install zc.buildout into it before running the buildout, something like this:

```
git clone mypackage
cd mypackage
python3 -m venv .
bin/pip install -U pip wheel zc.buildout
bin/buildout
bin/test
```

File: setup.py [hunk 24]
```diff
@@ -41,6 +47,7 @@
       description='A SQLAlchemy wrapper for Zope 2 and Zope 3',
       long_description=long_description,
       packages=find_packages('src'),
+      package_data={"": ["*.txt",]},
```
Comment (jordic) [inline setup.py:50]: perhaps this also should be removed

File: setup.py [hunk 25]
```diff
@@ -53,4 +60,11 @@
                         'zope.testing',
                         'zope.schema',
                        ],
-      extras_require=dict(test=['pysqlite', 'zope.testing']))
```
Comment (jordic) [inline setup.py:56]: revert this line also

**Comment by dataflake** [comment]
@jordic Now that your contributor agreement has been accepted you can create branches inside this repository. If you could move your changes from your fork to a branch in this repository other people could help a lot better.

**Comment by jordic** [comment]
Anything else missing?

**Comment by jordic** [comment]
> @jordic Now that your contributor agreement has been accepted you can create branches inside this repository. If you could move your changes from your fork to a branch in this repository other people could help a lot better.

I'm not pretty sure how to do this without loosing this helping conversation. Any hints? (I can do it on Products.SQLAlchemyDA).

File: src/z3c/sqlalchemy/util.py [hunk 26]
```diff
@@ -1,3 +1,4 @@
+from __future__ import print_function
```
Comment (mgedmin) [inline src/z3c/sqlalchemy/util.py:1]: The part about `__future__` imports being below the topmost block comment applies to this one too, but actually this import is not necessary in this file.  The only use of print() is at the very bottom, guarded by `if __name__ == '__main__':`, and it's a simple `print(expr)` that works in Python 2 just as well as in Python 3.

So maybe just drop the `__future__` import.

**Review by mgedmin — APPROVED** [review]
LGTM.

One minor comment.  I'd say no further review required if the only thing you change is remove that unnecessary `__future__` import from util.py

**Review by dataflake — APPROVED** [review]

**Comment by dataflake** [comment]
I want to apply more cleanups but it's too tedious to explain every single small thing when I could just do it myself once this is off the private fork and in this repository. Please merge.

**Comment by dataflake** [comment]
If you go to https://github.com/zopefoundation/z3c.sqlalchemy/commits/master you can look at those cleanups that bring the package more in line with current conventions, remove cruft and improve package management.

**Comment by jordic** [comment]
Uau, awesome!
